Fix parsing of fractional seconds.

igorpeshansky opened the pull request:
This addresses cases when fractional seconds aren't present, or when fractional seconds don't have exactly 9 digits.

bmoyles0117 commented:
What do you think about failing over to a secondary parser instead? If we can parse with nanoseconds, great, otherwise failover to without nanoseconds, or vice versa. It'd be nice to stick with the strptime utility instead of custom logic.

igorpeshansky commented:
There's no secondary parser — `strptime` does not support fractional seconds.
We could exclude seconds from the `strptime` format and then parse seconds and nanoseconds together as a `double` using `strtod`, which would avoid the whole length computation, but then we'd have to worry about funky formats like hex and floating point values like `"+inf"`, so the tradeoff is not obvious. I could try it and see, I suppose...

igorpeshansky commented:
I've tried it, and it would be fairly hard to detect cases like `"2018-03-03T01:23:00045.6789"` (i.e., 5 digits for the seconds) with the new approach (see the failing test). I'll leave the commit around in case you want to look at it, but I would lean towards removing it and sticking with the original implementation for now.

supriyagarg commented on src/time.cc:
```diff
@@ -58,28 +58,49 @@ std::string ToString(const std::chrono::system_clock::time_point& t) {
+    if (length > 9) {
+      // More digits than can be stored as nanoseconds.
+      // TODO: Should this round (std::lround)?
+      ns /= Exp10(length - 9);
```
round the value, rather than take floor?

igorpeshansky replied:
That would require changing the test as well. Which semantics do we want?

supriyagarg commented on src/time.cc:
```diff
@@ -59,51 +60,26 @@ std::string ToString(const std::chrono::system_clock::time_point& t) {
     return std::chrono::system_clock::time_point();
   }
+  tm.tm_sec = sec_i;
+  long ns = std::lround((seconds - sec_i) * 10000000000) / 10;
```
The new implementation looks good - much simpler. 
Please add a test for the rounding logic

bmoyles0117 commented on src/time.cc:
```diff
@@ -57,17 +62,31 @@ std::string ToString(const std::chrono::system_clock::time_point& t) {
-  if (zone <= end + 1 || *zone != 'Z' || *(zone+1) != '\0') {
-    // TODO
+  double seconds = std::strtod(end, &zone);
+  if (sec_i < 0 || sec_i != static_cast<long>(seconds)) {
```
Only minor note, I couldn't find a correlating unit test to this conditional statement. Every time there was a strange value in the place of the seconds, it was longer than 2 characters, that would have been caught by the previous condition. Might be overlooking the unit test that covers this case, just wanted to make a note.

igorpeshansky commented:
I've updated the new implementation to actually use the full time spec as well -- my one concern with the earlier implementation was that we cut off the seconds and parse them separately, which is now fixed. @bmoyles0117 @supriyagarg PTAL.

bmoyles0117 commented on src/time.cc:
```diff
@@ -57,17 +62,36 @@ std::string ToString(const std::chrono::system_clock::time_point& t) {
+    (void)std::strtol(end + 1, &zone, 10);
+    if (zone <= end + 1) {
+      // TODO: Missing nanoseconds.
+      return std::chrono::system_clock::time_point();
```
Is it not safe to assume 0 for nanoseconds?

igorpeshansky replied:
It depends on whether we treat a trailing decimal point with no digits after as valid input or an error. As I read the [spec](http://www.faqs.org/rfcs/rfc3339.html), a decimal point must be followed by at least one digit (search for `time-fraction`).

bmoyles0117 commented on src/time.cc:
```diff
@@ -57,17 +62,36 @@ std::string ToString(const std::chrono::system_clock::time_point& t) {
+    // TODO: Internal error.
     return std::chrono::system_clock::time_point();
   }
+  static_assert(sizeof(long) == 8, "long is too small");
```
When would this happen?

igorpeshansky replied:
For example with a 32-bit compiler. I'd rather fail compilation if someone ever tries to build for Amazon's 32-bit AMI or Windows than silently overflow.

igorpeshansky commented:
Rebased off `master`. Merging.
